nni/algorithms/compression/pytorch/pruning/transformer_pruner.py

```diff
@@ -0,0 +1,321 @@
+        Currently, the group inferred AttentionWeightDependency is limited to a set of four weights, with the first
+        three corresponding to Q_proj, K_proj, V_proj (in any order) and the last one being output_proj.
+        """
+        try:
+            module_graph = TorchModuleGraph(self.bound_model, self.dummy_input)
+            dependency_tracer = AttentionWeightDependency(traced_model=module_graph.trace)
+            self.attention_name_groups = dependency_tracer.dependency_sets
+            self.group_weights_by_name()
+
+        except Exception as e:
+            raise RuntimeError('Graph trace failed: please check dummy_input, or specify attention_name_groups.\n'
+                               'Exception message: ' + str(e))
+
+    def validate_weight_groups(self):
+        """
+        Sanity checks:
+            - Q, K, V projection weights in each groups must have the same shape
+            - output projection weight shape must match total hidden dimension (inferred from Q, K, V projection)
+            - Four weights in a group must have the same sparsity in their config
+            - If global_sort is specified, all weights must have the same sparsity
+            - head_hidden_dim must be a divisor of the output dimension of the projection weights (i.e., the resulting
+              head number must be an integer)
+        """
+        errmsg = 'Attention weight group sanity check not passed'
+        sparsity = None
+        for group in self.masking_groups:
+            # allow empty groups - may be caused by config list filtering
+            if len(group) == 0:
+                continue
+            assert len(group) == 4, errmsg + ': each group must have four weights'
+            assert group[0].module.weight.size() == group[1].module.weight.size() and \
+                group[1].module.weight.size() == group[2].module.weight.size(), \
+                errmsg + ': the dimensions of Q, K, V projection matrices must be the same '
+            assert group[0].module.weight.size()[0] == group[3].module.weight.size()[1], \
+                errmsg + ': the dimension of attention results must match with input for output projection'
+            assert group[0].config['sparsity'] == group[1].config['sparsity'] == \
+                   group[2].config['sparsity'] == group[3].config['sparsity'], \
+                errmsg + ': the sparsity of matrices in the same layer must be the same'
+            if sparsity is None:
+                sparsity = group[0].config['sparsity']
+            if self.global_sort:
+                assert sparsity == group[0].config['sparsity'], \
+                    errmsg + ': for global_sort=True, the sparsity for all modules must be the same'
+            t = group[0].module.weight.size(0) / self.head_hidden_dim
+            assert t % 1 == 0, errmsg + ': head_hidden_dim must be a divisor of the output dimension of the ' \
```

Comment:
?

Reply:
That was meant to check that `t` is an integer. Not sure if there are better ways to implement that

Reply:
assert group[0].module.weight.size(0) % self.head_hidden_dim == 0

Reply:
fixed